lib/cobra_commander/component_tree/combined_component_tree.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require "oj"
+
+module CobraCommander
+  class ComponentTree
+    # Represents a tree of all component dependencies
+    class CombinedComponentTree
+      def initialize(name, path, ancestry = Set.new)
+        @name = name
+        @root_path = path
+        @ancestry = ancestry
+      end
+
+      def to_h
+        {
+          name: @name,
+          path: @root_path,
+          ancestry: @ancestry,
+          dependencies: dependencies.map(&method(:dep_representation)),
+        }
+      end
+
+      def node_tree
+        @node_tree ||= NodeComponentTree.new(UMBRELLA_APP_NAME, @root_path).to_h
+      end
+
+      def gem_tree
+        @gem_tree ||= GemComponentTree.new(UMBRELLA_APP_NAME, @root_path).to_h
+      end
+
+      def js_format(deps)
+        linked_deps = deps.select { |_, v| v.start_with? "link:" }
+        linked_deps.map do |_, v|
+          relational_path = v.split("link:")[1]
+          dep_name = relational_path.split("/")[-1]
+          { name: dep_name, path: relational_path }
+        end
+      end
+
+      def gemfile_path
+        File.join(@root_path, "Gemfile")
+      end
+
+      def gemfile_lock_path
+        File.join(@root_path, "Gemfile.lock")
+      end
+
+      def bundler_definition
+        ::Bundler::Definition.build(gemfile_path, gemfile_lock_path, nil)
+      end
+
+      def dependencies
+        @deps ||= begin
+          deps = []
+          deps << component_dependencies if File.exist?(File.join(@root_path, "Gemfile"))
```

Comment:
Worth moving these conditionals into the respective query methods and just return an empty collection? I'm not keen on having to check the safety of calling the query methods.